### ISSUE TITLE
Create Intellij Idea plugin for the nu language

### ISSUE THREAD

**Issue by mmerdes** [nushell/nushell#7160]
### Related problem

In order to create programs rather than one-line-scripts, IDE support is very helpful.
It seems that there are a number of editor/IDE plugins, but Idea is not supported.

### Describe the solution you'd like

Provide a plugin for nu to work natively within Idea.

### Describe alternatives you've considered

I am aware of the existing alternatives. 
Providing a plugin for Idea would help nu to better reach the potentially huge audience of enterprise developers.

### Additional context and details

As far as I know such a plugin can be used in different Jetbrains IDEs which offer support for many programming languages.

**Comment by fdncred** [comment]
Are there specifications somewhere on what these idea plugins need to look like?

**Comment by mmerdes** [comment]
Thank you @fdncred for your fast reaction.
The 'Intellij Platform' allows for building all kinds of plugins.
There is also a dedicated description for creating a plugin to support a new programming language:
https://plugins.jetbrains.com/docs/intellij/custom-language-support.html

This includes a detailed description on how to implement support for a custom language such as nu - 
starting from lexer and parser, syntax highlighting and code completion to refactoring and other advanced features.

Hope this helps.

**Comment by fdncred** [comment]
Good luck to whomever wants to do this. Java is a hard pass for me. Sorry. But that is good documentation. So, for someone who is familiar with java, it may not be that tricky. Probably the biggest missing piece in nushell today, as it relates to IDE support, is that we do not have an LSP. So, I'd guess that the best we can do right now is to just do something with syntax highlighting.

**Comment by dbuch** [comment]
> LSPs are mainly useful for Microsoft-related products

I'm sure thats wrong (though microsoft seem to be the top maintainer). LSP is a protocol (Language server protocol). So it only dictates the way the client and server communicates. It's built on JSON-RPC. @mmerdes you can read more on the [official lsp page](https://microsoft.github.io/language-server-protocol/)

_In theory_ the server _could_ be implemented in nu-language. Which in it self sound cool. However I dont think NU is ready for that. One missing piece would probably be background processing capabilities (threads/async). Side note: One of the popular lsp implementations for lua is actually written in lua: https://github.com/sumneko/lua-language-server

I think the most reasonable would be to implement the server in rust which can link to nu-parser crate or so. That could (I believe) resolve symbols for autocompletion and parse failures for diagnostics or so.

Once that is done, **any** client like nvim/vim, vscode, helix and apparently fleet. Will be able to equally develop nu code.

It would be super cool and I agree - it's needed. But it require alot of work (I think?) and planning.
On top of my head:

- What does a "document symbol" even mean in the Nu context?
- What and how should code actions do, on what basis?

As a MVP I think diagnostics and internal command completion would be a good start.

Above is just 50 cents from a "drop-by" user.

**Comment by it-a-me** [comment]
I'm not sure how feasible it would be, but I'd love to have the lsp to be tied to the nu binary itself, possibly behind a cargo flag.  Elvish, Julia, R, dart, and bass already do.  The major advantage is that regardless of possible breaking changes, the lsp always matches the current nu version and provides the exact same error messages as running the script would.  Another minor advantage is that it helps the lsp get packaged/shiped to end users without shiping duplicates of the complied nu parsing code

**Comment by NiceGuyIT** [comment]
I too would like this. Version [0.79.0](https://www.nushell.sh/blog/2023-04-25-nushell_0_79.html#enhanced-ide-support-in-our-vs-code-extension-jt-fdncred) introduced a [VSCode extension](https://github.com/nushell/vscode-nushell-lang/issues/117) that uses an LSP. Unfortunately it's written in JavaScript and specific to the extension. There's an issue to create a [standalone LSP Server in Rust](https://github.com/nushell/vscode-nushell-lang/issues/117). That would make this task easier.

**Comment by fdncred** [comment]
Stay tuned, we have plans.
